File: backend/src/routers/movie_assets.py
from fastapi import (
    APIRouter,
    Depends,
    Request,
    Query,
    Path,
    Body,
    UploadFile,
    File,
    Form,
    HTTPException,
)
from typing import List, Optional
import os
from urllib.parse import urlparse

from src.core.dependencies import get_current_user
from src.models import AssetStoreType, AssetType, AssetUpdate, AssetCreate
from src.routers.schemas.asset import (
    AssetReadResponseSchema,
    AssetPageResultResponseSchema,
    AssetCreateRequestSchema,
    AssetUpdateRequestSchema,
)
from src.routers.schemas.user_asset import UserAssetIdsRequestSchema
from src.services import MovieAssetService
from src.core.handler import router_handler
from src.core.idempotency import idempotent
from config.logging import get_router_logger

router = APIRouter(prefix="/assets", tags=["Movies"])
asset_service = MovieAssetService()
logger = get_router_logger("movie_assets")

@router.get("/", response_model=AssetPageResultResponseSchema)
@router_handler(action="list_all_assets")
async def list_all_assets(
    request: Request,
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=200, description="每页数量"),
    current_user=Depends(get_current_user),
):
    return await asset_service.list_all_assets_page(
        page=page, size=size, current_user=current_user
    )


# 查找（Query）
@router.get("/recycle-bin", response_model=AssetPageResultResponseSchema)
@router_handler(action="list_recycle_bin_assets")
async def list_recycle_bin_assets(
    request: Request,
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=200, description="每页数量"),
    current_user=Depends(get_current_user),
):
    return await asset_service.list_recycle_bin_assets(
        page=page, size=size, current_user=current_user
    )


@router.get("/movies/{movie_id}", response_model=AssetPageResultResponseSchema)
@router_handler(action="list_movie_assets")
async def list_movie_assets(
    request: Request,
    movie_id: str = Path(..., description="影片ID"),
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=200, description="每页数量"),
    current_user=Depends(get_current_user),
):
    return await asset_service.list_movie_assets_page(
        movie_id, page=page, size=size, current_user=current_user
    )

@router.get("/{asset_id}", response_model=AssetReadResponseSchema)
@router_handler(action="get_asset")
async def get_asset(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    current_user=Depends(get_current_user),
):
    return await asset_service.get_asset(
        asset_id, current_user=current_user
    )


# 修改（Update）
@router.patch("/{asset_id}", response_model=AssetReadResponseSchema)
@router_handler(action="update_movie_asset")
async def update_movie_asset(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    patch: AssetUpdateRequestSchema = Body(...),
    current_user=Depends(get_current_user),
):
    model_patch = AssetUpdate(**patch.model_dump(exclude_unset=True))
    return await asset_service.update_movie_asset(
        asset_id,
        model_patch,
        current_user=current_user,
    )

# 删除（Delete）
@router.delete("/bulk", response_model=dict)
@router_handler(action="delete_movie_assets_bulk")
async def delete_movie_assets(
    request: Request,
    body: UserAssetIdsRequestSchema = Body(..., description="资产ID列表"),
    soft_delete: bool = Query(True, description="是否软删除"),
    current_user=Depends(get_current_user),
):
    asset_ids = body.asset_ids
    success = await asset_service.delete_movie_assets(
        asset_ids,
        current_user=current_user,
        soft_delete=soft_delete,
    )
    failed = max(len(asset_ids) - success, 0)
    message = "删除完成" if failed == 0 else "部分资产删除失败"
    return {"message": message, "success": success, "failed": failed}

@router.delete("/{asset_id}", response_model=dict)
@router_handler(action="delete_movie_asset")
async def delete_movie_asset(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    soft_delete: bool = Query(True, description="是否软删除"),
    current_user=Depends(get_current_user),
):
    ok = await asset_service.delete_movie_asset(
        asset_id,
        current_user=current_user,
        soft_delete=soft_delete,
    )
    return {"message": "删除完成" if ok else "删除失败", "success": 1 if ok else 0, "failed": 0 if ok else 1}


# 重建（Rebuild / Restore）
@router.post("/restore", response_model=List[AssetReadResponseSchema])
@router_handler(action="restore_movie_assets_bulk")
async def restore_movie_assets(
    request: Request,
    body: UserAssetIdsRequestSchema = Body(..., description="资产ID列表"),
    current_user=Depends(get_current_user),
):
    asset_ids = body.asset_ids
    restored = await asset_service.restore_movie_assets(
        asset_ids,
        current_user=current_user,
    )
    return restored

# 新增（Create）
@router.post("/{movie_id}", response_model=dict)
@router_handler(action="create_movie_asset")
@idempotent()
async def create_movie_asset(
    request: Request,
    movie_id: str = Path(..., description="影片ID"),
    library_id: str = Form(..., description="媒体库ID"),
    type: AssetType = Form(..., description="资产类型"),
    name: Optional[str] = Form(None, description="资产名称"),
    store_type: Optional[str] = Form("Local", description="资产存储"),  # TODO 后期存储扩展
    file: Optional[UploadFile] = File(None, description="媒体文件"),
    url: Optional[str] = Form(None, description="媒体文件可访问的 URL（http/https）"),
    local_path: Optional[str] = Form(None, description="服务端源文件绝对路径"),
    source_ext: Optional[str] = Form(None, description="源文件扩展名"),
    current_user=Depends(get_current_user),
):
    sources = [src for src in (file, url, local_path) if src is not None]
    if len(sources) != 1:
        raise HTTPException(status_code=400, detail="请提供且仅提供其一：file / url / local_path")

    if url is not None:
        parsed = urlparse(url)
        if parsed.scheme not in ("http", "https") or not parsed.netloc:
            raise HTTPException(status_code=400, detail="非法 URL：仅支持 http/https")
    if local_path is not None:
        if not os.path.isabs(local_path) or not os.path.exists(local_path):
            raise HTTPException(status_code=400, detail="源文件路径不存在或不可访问（需绝对路径）")

    # 路由层：使用 schemas 入参并转换为 Model 层
    create_req = AssetCreateRequestSchema(
        movie_id=movie_id,
        type=type,
        name=name,
    ).model_dump()
    create_req['path'] = 'temp'
    create_req['library_id'] = library_id
    payload = AssetCreate(**create_req)
    ctx = {"file": file, "url": url, "local_path": local_path, "source_ext": source_ext}
    asset_model, task_info = await asset_service.upload_movie_asset(
        payload,
        context=ctx,
        current_user=current_user,
    )
    return {"asset": asset_model, "tasks": task_info}

@router.post("/thumbnails/sign", response_model=List[str])
@router_handler(action="get_asset_thumbnails_signed")
async def get_asset_thumbnails_signed(
    request: Request,
    body: UserAssetIdsRequestSchema = Body(..., description="资产ID列表"),
    current_user=Depends(get_current_user),
):
    return await asset_service.list_asset_thumbnails_signed(
        body.asset_ids,
        current_user=current_user,
    )

@router.get("/{asset_id}/file")
@router_handler(action="get_movie_asset_file")
async def get_movie_asset_file(
    request: Request,
    asset_id: str = Path(..., description="资产ID"),
    transcode: bool = Query(False, description="是否实时转码"),
    start: Optional[float] = Query(None, ge=0, description="转码起始秒"),
    duration: Optional[float] = Query(None, ge=1, description="转码时长（秒）"),
    target_bitrate_kbps: Optional[int] = Query(None, ge=64, le=100000, description="目标视频码率（kbps），与分辨率二选一"),
    target_resolution: Optional[str] = Query(None, description="目标分辨率，例如 1280x720；与码率二选一"),
    current_user=Depends(get_current_user),
):
    return await asset_service.get_asset_file(
        asset_id,
        transcode=transcode,
        start=start,
        duration=duration,
        target_bitrate_kbps=target_bitrate_kbps,
        target_resolution=target_resolution,
        current_user=current_user,
    )

File: frontend/src/api/movie_assets.ts
import { http, resolveBaseURL, resolveMediaURL } from "./http";
import type {
  AssetType,
  AssetRead,
  AssetPageResult,
  AssetUpdateRequestSchema,
} from "../types/asset";

/**
 * 全局资产列表：GET /assets
 */
export async function getGlobalAssets(
  token: string,
  params: { page?: number; size?: number } = {},
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetPageResult> {
  const qp = new URLSearchParams();
  if (params.page !== undefined) qp.set("page", String(params.page));
  if (params.size !== undefined) qp.set("size", String(params.size));
  return http.get<AssetPageResult>("/assets/", {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
}

/**
 * 资产列表：GET /assets/movies/{movie_id}
 */
export async function getMovieAssets(
  token: string,
  movieId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetPageResult> {
  return http.get<AssetPageResult>(`/assets/movies/${movieId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 回收站列表：GET /assets/recycle-bin
 */
export async function listRecycleBinAssets(
  token: string,
  params: { page?: number; size?: number } = {},
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetPageResult> {
  const qp = new URLSearchParams();
  if (params.page !== undefined) qp.set("page", String(params.page));
  if (params.size !== undefined) qp.set("size", String(params.size));
  return http.get<AssetPageResult>("/assets/recycle-bin", {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
}

/**
 * 通过 URL 导入资产：POST /assets/{movie_id}
 * FormData: type, url, name?
 * 返回响应字典中的 asset 字段
 */
export async function importMovieAssetFromUrl(
  token: string,
  movieId: string,
  libraryId: string,
  params: { type: AssetType; url: string; name?: string | null },
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetRead> {
  const form = new FormData();
  form.append("type", params.type as unknown as string);
  form.append("url", params.url);
  if (params.name) form.append("name", params.name);
  form.append("library_id", libraryId);
  const res = await http.post<{ asset: AssetRead; tasks: unknown }>(`/assets/${movieId}`, form, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return res.asset;
}

/**
 * 上传资产文件：POST /assets/{movie_id}
 * FormData: type, name?, file
 * 返回响应字典中的 asset 字段
 */
export async function uploadMovieAssetFile(
  token: string,
  movieId: string,
  libraryId: string,
  params: { type: AssetType; name?: string | null },
  file: File | Blob,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetRead> {
  const form = new FormData();
  form.append("type", params.type as unknown as string);
  if (params.name) form.append("name", params.name);
  form.append("file", file);
  form.append("library_id", libraryId);
  const res = await http.post<{ asset: AssetRead; tasks: unknown }>(`/assets/${movieId}`, form, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return res.asset;
}

/**
 * 服务端本地路径复制上传：POST /assets/{movie_id}
 * FormData: type, name?, local_path
 * 返回响应字典中的 asset 字段
 */
export async function importMovieAssetFromLocal(
  token: string,
  movieId: string,
  libraryId: string,
  params: { type: AssetType; src_path: string; name?: string | null },
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetRead> {
  const form = new FormData();
  form.append("type", params.type as unknown as string);
  form.append("local_path", params.src_path);
  if (params.name) form.append("name", params.name);
  form.append("library_id", libraryId);
  const res = await http.post<{ asset: AssetRead; tasks: unknown }>(`/assets/${movieId}`, form, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
  return res.asset;
}

/**
 * 更新资产：PATCH /assets/{asset_id}
 */
export async function updateMovieAsset(
  token: string,
  movieId: string,
  assetId: string,
  patch: AssetUpdateRequestSchema,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetRead> {
  return http.patch<AssetRead>(`/assets/${assetId}`, patch, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 删除资产：DELETE /assets/{asset_id}?soft_delete=
 */
export async function deleteMovieAsset(
  token: string,
  movieId: string,
  assetId: string,
  softDelete = true,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<Record<string, unknown>> {
  const qp = new URLSearchParams();
  qp.set("soft_delete", String(softDelete));
  return http.delete<Record<string, unknown>>(`/assets/${assetId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
}

/**
 * 获取单个资产详情：GET /assets/{asset_id}
 */
export async function getMovieAsset(
  token: string,
  assetId: string,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetRead> {
  return http.get<AssetRead>(`/assets/${assetId}`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 获取资产文件签名URL：GET /assets/{asset_id}/file
 * 返回字符串URL，适用于 <video src> 直接使用（无额外请求头）
 */
export async function getMovieAssetFile(
  token: string,
  assetId: string,
  params?: {
    transcode?: boolean;
    start?: number;
    duration?: number;
    target_bitrate_kbps?: number;
    target_resolution?: string;
  },
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<string> {
  const qp = new URLSearchParams();
  if (params?.transcode !== undefined) qp.set("transcode", String(params.transcode));
  if (params?.start !== undefined) qp.set("start", String(params.start));
  if (params?.duration !== undefined) qp.set("duration", String(params.duration));
  if (params?.target_bitrate_kbps !== undefined) qp.set("target_bitrate_kbps", String(params.target_bitrate_kbps));
  if (params?.target_resolution) qp.set("target_resolution", params.target_resolution);
  return http.get<string>(`/assets/${assetId}/file`, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
    query: qp,
  });
}

/**
 * 批量删除资产：DELETE /assets/bulk?soft_delete=
 * Body: { asset_ids: string[] }
 */
export async function deleteMovieAssetsBulk(
  token: string,
  assetIds: string | string[],
  softDelete = true,
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<{ message: string; success: number; failed: number }> {
  const idsArray = Array.isArray(assetIds)
    ? assetIds
    : assetIds.split(",").map((s) => s.trim()).filter(Boolean);
  const qp = new URLSearchParams();
  qp.set("soft_delete", String(softDelete));
  const body = { asset_ids: idsArray };
  return http.deleteWithBody<{ message: string; success: number; failed: number }>(
    "/assets/bulk",
    body,
    {
      token,
      baseURL: options?.baseURL,
      signal: options?.signal,
      query: qp,
    }
  );
}

/**
 * 批量恢复资产：POST /assets/restore
 * Body: { asset_ids: string[] }
 */
export async function restoreMovieAssets(
  token: string,
  assetIds: string | string[],
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<AssetRead[]> {
  const idsArray = Array.isArray(assetIds)
    ? assetIds
    : assetIds.split(",").map((s) => s.trim()).filter(Boolean);
  const body = { asset_ids: idsArray };
  return http.post<AssetRead[]>("/assets/restore", body, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 获取资产缩略图签名：POST /assets/thumbnails/sign
 * Body: { asset_ids: string[] }
 * 返回签名 URL 列表
 */
export async function getAssetThumbnailsSigned(
  token: string,
  assetIds: string | string[],
  options?: { baseURL?: string; signal?: AbortSignal }
): Promise<string[]> {
  const idsArray = Array.isArray(assetIds)
    ? assetIds
    : assetIds.split(",").map((s) => s.trim()).filter(Boolean);
  const body = { asset_ids: idsArray };
  return http.post<string[]>("/assets/thumbnails/sign", body, {
    token,
    baseURL: options?.baseURL,
    signal: options?.signal,
  });
}

/**
 * 构造资产文件下载/播放 URL：GET /assets/{asset_id}/file
 * 用于 <video src> 或下载链接，不直接发起请求。
 */
export function buildAssetFileURL(
  assetId: string,
  params?: {
    transcode?: boolean;
    start?: number;
    duration?: number;
    target_bitrate_kbps?: number;
    target_resolution?: string;
  },
  token?: string,
  baseURL?: string
): string {
  const qp = new URLSearchParams();
  if (token) qp.set("token", token);
  if (params?.transcode != null) qp.set("transcode", String(params.transcode));
  if (params?.start != null) qp.set("start", String(params.start));
  if (params?.duration != null) qp.set("duration", String(params.duration));
  if (params?.target_bitrate_kbps != null)
    qp.set("target_bitrate_kbps", String(params.target_bitrate_kbps));
  if (params?.target_resolution) qp.set("target_resolution", params.target_resolution);
  const base = resolveMediaURL(baseURL);
  const qs = qp.toString();
  const path = `/assets/${assetId}/file${qs ? `?${qs}` : ""}`;
  return `${base}${path}`;
}